### PR TITLE
Implement first-class support for the new Buffers in transports

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
Our transports should be able to directly read and write buffers with the new API, without going through a ByteBuf intermediary object.

Modification:
Implement reading and writing of Buffers, both for streaming channels, and for datagrams.

Result:
It is now in principle possible to use the new buffer API end-to-end.

### CODE REVIEW

File: buffer/src/main/java/io/netty/buffer/api/Buffer.java [hunk 1]
```diff
@@ -212,6 +212,14 @@
      */
     boolean readOnly();
 
+    /**
+     * Queries if this buffer is backed by native memory, or not.
+     *
+     * @return {@code true} if this buffer is backed by native, off-heap, memory. Otherwise, {@code false}, if this
+     * buffer is backed by on-heap memory.
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/api/Buffer.java:219]: @chrisvest I  just wonder if `isDirect()` could be false and it is still non on-heap in theory. Like for example an empty buffer.

File: buffer/src/main/java/io/netty/buffer/api/DefaultCompositeBuffer.java [hunk 2]
```diff
@@ -360,6 +360,16 @@ public boolean readOnly() {
...
+    @Override
+    public boolean isDirect() {
+        // A composite buffer is direct, if all components are direct.
+        boolean direct = true;
+        for (Buffer buf : bufs) {
+            direct &= buf.isDirect();
+        }
+        return direct;
+    }
+
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/api/DefaultCompositeBuffer.java:372]: There is no need to continue looping if we found one `non direct` buffer:

```suggestion
    public boolean isDirect() {
        // A composite buffer is direct, if all components are direct.
        for (Buffer buf : bufs) {
            if (!buf.isDirect()) {
                return false;
            }
        }
        return true;
    }

```

File: buffer/src/main/java/io/netty/buffer/api/internal/LifecycleTracer.java [hunk 3]
```diff
@@ -37,7 +37,7 @@
 @UnstableApi
 public abstract class LifecycleTracer {
     static final boolean lifecycleTracingEnabled =
-            SystemPropertyUtil.getBoolean("io.netty.buffer.lifecycleTracingEnabled", false);
+            SystemPropertyUtil.getBoolean("io.netty.buffer.lifecycleTracingEnabled", true);
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty/buffer/api/internal/LifecycleTracer.java:40]: revert ?

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: buffer/src/main/java/io/netty/buffer/api/Buffer.java (hunk 1 above)
Comment (chrisvest) [reply buffer/src/main/java/io/netty/buffer/api/Buffer.java:219]: Empty buffers are not special in the new API. You can allocate empty off-heap buffers, and empty on-heap buffers.

**Comment by normanmaurer** [comment]
@chrisvest ping me once I should have a look again

File: codec/src/test/java/io/netty/handler/codec/DatagramPacketEncoderTest.java [hunk 4]
```diff
@@ -82,11 +82,7 @@ public void testUnmatchedMessageType() {
                 new DefaultAddressedEnvelope<>(1L, recipient, sender);
         assertTrue(channel.writeOutbound(envelope));
         DefaultAddressedEnvelope<Long, InetSocketAddress> output = channel.readOutbound();
-        try {
-            assertSame(envelope, output);
-        } finally {
-            output.release();
-        }
+        assertSame(envelope, output);
```
Comment (normanmaurer) [inline codec/src/test/java/io/netty/handler/codec/DatagramPacketEncoderTest.java:85]: don't we still need to "close" the output ?

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java [hunk 5]
```diff
@@ -142,15 +146,107 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
...
+                                    aggregator.writeBytes(buf);
+                                }
+                            }
+                        }
+
+                        @Override
+                        public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                            // IOException is fine as it will also close the channel and may just be a connection reset.
+                            if (!(cause instanceof IOException)) {
+                                clientReceived.set(cause);
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java:195]: Should we also close the `aggregator` in this case ?

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java [hunk 6]
```diff
@@ -282,6 +378,139 @@ public void channelInactive(ChannelHandlerContext ctx) throws Exception {
...
+                                latch.countDown();
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java:466]: should we also close the aggregator here ?

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/DatagramConnectNotExistsTest.java [hunk 7]
```diff
@@ -45,6 +46,38 @@
...
+            }
+        });
+        Future<Channel> future = cb.connect(NetUtil.LOCALHOST, SocketTestPermutation.BAD_PORT);
+        Channel datagramChannel = null;
+        try {
+            datagramChannel = future.get();
+            assertTrue(datagramChannel.isActive());
+            datagramChannel.writeAndFlush(
+                    Unpooled.copiedBuffer("test", CharsetUtil.US_ASCII)).syncUninterruptibly();
+            assertTrue(promise.asFuture().syncUninterruptibly().getNow() instanceof PortUnreachableException);
```
Comment (normanmaurer) [inline testsuite/src/main/java/io/netty/testsuite/transport/socket/DatagramConnectNotExistsTest.java:73]: nit: use `assertThat(...)` for better messages

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: codec/src/test/java/io/netty/handler/codec/DatagramPacketEncoderTest.java (hunk 4 above)
Comment (chrisvest) [reply codec/src/test/java/io/netty/handler/codec/DatagramPacketEncoderTest.java:85]: Not in this case. It's a `DefaultAddressedEnvelope` that just contains a `Long`. Nothing to close.

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java (hunk 5 above)
Comment (chrisvest) [reply testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java:195]: Does `channelInactive` not take care of it regardless?

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java (hunk 6 above)
Comment (chrisvest) [reply testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java:466]: Does `channelInactive` not take care of it regardless?

**Comment by chrisvest** [comment]
@normanmaurer This is ready for a proper review now

File: transport/src/main/java/io/netty/channel/DefaultChannelHandlerContext.java [hunk 8]
```diff
@@ -351,13 +352,21 @@ public ChannelHandlerContext fireChannelRead(final Object msg) {
...
         if (ctx == null) {
-            ReferenceCountUtil.release(msg);
+            closeOrRelease(msg);
             notifyHandlerRemovedAlready();
             return;
         }
         ctx.invokeChannelRead(msg);
     }
 
+    private static void closeOrRelease(Object msg) {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/DefaultChannelHandlerContext.java:362]: I saw this method a lot of times... Consider factoring out to a static helper method.

File: transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java [hunk 9]
```diff
@@ -191,6 +211,14 @@ public final void read() {
                 }
             }
         }
+
+        private void closeOrRelease(Object obj) {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java:215]: See above

File: transport/src/main/java/io/netty/channel/socket/nio/NioDatagramChannel.java [hunk 10]
```diff
@@ -230,24 +238,56 @@ protected void doClose() throws Exception {
 
     @Override
     protected int doReadMessages(List<Object> buf) throws Exception {
-        DatagramChannel ch = javaChannel();
         DatagramChannelConfig config = config();
         RecvBufferAllocator.Handle allocHandle = unsafe().recvBufAllocHandle();
 
+        if (config.getRecvBufferAllocatorUseBuffer()) {
+            return doReadBufferMessages(allocHandle, buf);
+        } else {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/socket/nio/NioDatagramChannel.java:246]: nit: you can remove the else

File: transport/src/test/java/io/netty/channel/local/LocalChannelTest.java [hunk 11]
```diff
@@ -1352,7 +1352,6 @@ public void messageReceived(ChannelHandlerContext ctx, Buffer buffer) throws Exc
                                     // Fill the ChannelOutboundBuffer with multiple buffers
                                     ctx.write(buffer.readSplit(1));
                                 }
-                                buffer.close();
```
Comment (normanmaurer) [inline transport/src/test/java/io/netty/channel/local/LocalChannelTest.java:1355]: Why this change ?

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java [hunk 12]
```diff
@@ -291,6 +520,16 @@ private static ByteBuf newCompositeBuffer(ByteBufAllocator alloc) {
         return compositeByteBuf;
     }
 
+    private static Buffer newCompositeBuffer(BufferAllocator alloc) {
+        CompositeBuffer compositeBuffer = CompositeBuffer.compose(
+                alloc,
+                alloc.allocate(4).writeInt(100).send(),
+                alloc.allocate(8).writeLong(123).send(),
+                alloc.allocate(8).writeLong(456).send());
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java:528]: As far as I can see, those methods are only used in testing. And that doesn't strike me as a particularly strong case for adding them.

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/WriteBeforeRegisteredTest.java [hunk 13]
```diff
@@ -28,6 +29,28 @@
...
+
+    public void testWriteBeforeConnectByteBuf(Bootstrap cb) throws Throwable {
+        TestHandler h = new TestHandler();
+        SocketChannel ch = null;
+        try {
+            cb.handler(h);
+            ch = (SocketChannel) cb.createUnregistered();
+            ch.writeAndFlush(Unpooled.wrappedBuffer(new byte[] { 1 }));
+            ch.register().sync();
+            ch.connect(newSocketAddress());
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply testsuite/src/main/java/io/netty/testsuite/transport/socket/WriteBeforeRegisteredTest.java:46]: There's nothing listening on that address. This test was always kind of weird. See the previous change: https://github.com/netty/netty/commit/6b11f7fbc24c0bcfad86aac323b8d418672f88ff#diff-8cc43f43a168da2284124b1c94b57b381bc45e42f5a0d8a9c94da2a0308a598e

File: transport/src/test/java/io/netty/channel/local/LocalChannelTest.java (hunk 11 above)
Comment (chrisvest) [reply transport/src/test/java/io/netty/channel/local/LocalChannelTest.java:1355]: The `SimpleChannelInboundHandler` releases/closes the message for us, and that would produce an exception if we already closed it here.
This doesn't impact the test, though. It just logs the exception.

**Comment by chrisvest** [comment]
@normanmaurer Addressed/replied to comments. Please take a look.

File: testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (normanmaurer) [reply testsuite/src/main/java/io/netty/testsuite/transport/socket/CompositeBufferGatheringWriteTest.java:542]: fair enough

File: transport/src/test/java/io/netty/channel/local/LocalChannelTest.java (hunk 11 above)
Comment (normanmaurer) [reply transport/src/test/java/io/netty/channel/local/LocalChannelTest.java:1355]: got it...

**Review by normanmaurer — CHANGES_REQUESTED** [review]
almost there

File: transport-classes-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java [hunk 14]
```diff
@@ -135,6 +139,41 @@ private int writeBytes(ChannelOutboundBuffer in, ByteBuf buf) throws Exception {
...
+
+        int readableComponents = buf.countReadableComponents();
+        if (readableComponents == 1) {
+            return doWriteBytes(in, buf);
+        } else {
+            ByteBuffer[] nioBuffers = new ByteBuffer[readableComponents];
+            buf.forEachReadable(0, (index, component) -> {
+                nioBuffers[index] = component.readableBuffer();
+                return true;
+            });
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply transport-classes-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java:170]: Would you do the IO inside the function? I suppose that would have to be the case to avoid escape. I think we can do this.

File: buffer/src/main/java/io/netty/buffer/api/Resource.java [hunk 15]
```diff
@@ -66,4 +68,33 @@
...
+            try {
+                ((AutoCloseable) obj).close();
+            } catch (RuntimeException re) {
+                throw re;
+            } catch (Exception e) {
+                throw new RuntimeException("Exception from closing object", e);
+            }
+        } else if (obj instanceof ReferenceCounted) {
+            ReferenceCounted rc = (ReferenceCounted) obj;
+            if (rc.refCnt() > 0) {
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply buffer/src/main/java/io/netty/buffer/api/Resource.java:95]: Yeah, good point.

File: transport-classes-kqueue/src/main/java/io/netty/channel/kqueue/AbstractKQueueStreamChannel.java [hunk 16]
```diff
@@ -353,7 +393,6 @@ private int doWriteMultiple(ChannelOutboundBuffer in) throws Exception {
         in.forEachFlushedMessage(array);
 
         if (array.count() >= 1) {
-            // TODO: Handle the case where cnt == 1 specially.
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply transport-classes-kqueue/src/main/java/io/netty/channel/kqueue/AbstractKQueueStreamChannel.java:356]: The `doWriteMultiple` method is only called when `ChannelOutboundBuffer.size()` is greater than 1, so `array.count()` can only be 1 if there were empty buffers that got skipped, which doesn't sound like a very common or important case to optimise.

File: transport-classes-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (normanmaurer) [reply transport-classes-epoll/src/main/java/io/netty/channel/epoll/AbstractEpollStreamChannel.java:170]: yeah... The only "downside" I can see is that we would need to do "boxing" / "unboxing" or add extra methods that take and `ToIntFunction` / `ToLongFunction` etc . I guess that can easily be done tho... Let me do a PR to add what I suggest and we can change the code to make use of it later on.
